File: 1122-relative-sort-array/custom-sort-with-inline-composite-key/cpp/solution.cpp
class Solution {
public:
    vector<int> relativeSortArray(vector<int>& arr1, vector<int>& arr2) {
        unordered_map<int, int> order;
        for (int idx = 0; idx < arr2.size(); idx++) 
            order[arr2[idx]] = idx;

        const int INF = numeric_limits<int>::max();

        sort(arr1.begin(), arr1.end(), [&](int a, int b) {
            auto itA = order.find(a);
            auto itB = order.find(b);
            // If in arr2, use its index, otherwise, treat as INF
            int rankA = (itA != order.end()) ? itA->second : INF;
            int rankB = (itB != order.end()) ? itB->second : INF;

            // Primary sort by arr2 index, secondary sort by value (asc).
            return (rankA != rankB) ? rankA < rankB : a < b;    
        });
        return arr1;     
    }
};

File: 1122-relative-sort-array/custom-sort-with-inline-composite-key/java/solution.java
class Solution {
    public int[] relativeSortArray(int[] arr1, int[] arr2) {
        // Map for rank lookups
        Map<Integer, Integer> order = new HashMap<>();  
        for (int idx = 0; idx < arr2.length; idx++) 
            order.put(arr2[idx], idx);
        
        // Explicit Boxing (Required for custom Comparator support)
        Integer[] boxed = new Integer[arr1.length]; 
        for (int idx = 0; idx < arr1.length; idx++) 
            boxed[idx] = arr1[idx];

        // Inline Custom Sort
        Arrays.sort(boxed, (a, b) -> {
            int rankA = order.getOrDefault(a, Integer.MAX_VALUE);
            int rankB = order.getOrDefault(b, Integer.MAX_VALUE);
        return (rankA != rankB) ? rankA - rankB : a - b;
        });

        // Explicit Unboxing
        for (int idx = 0; idx < arr1.length; idx++)
            arr1[idx] = boxed[idx];
        
        return arr1;
    }
}
